folly/CPortability.h

```diff
@@ -53,4 +53,725 @@
 # define FOLLY_DISABLE_ADDRESS_SANITIZER
 #endif
 
+
+#ifdef _MSC_VER
+#include <sys/stat.h>
+#define S_IXUSR 0
+#define S_IWUSR _S_IWRITE
+#define S_IRUSR _S_IREAD
+
+#define S_IXGRP S_IXUSR
+#define S_IWGRP S_IWUSR
+#define S_IRGRP S_IRUSR
+#define S_IXOTH S_IXUSR
+#define S_IWOTH S_IWUSR
+#define S_IROTH S_IRUSR
+#endif
+
+
+#ifdef _MSC_VER
+
+#ifndef __STDC__
+#define __STDC__ 1
+#include <io.h>
+#undef __STDC__
+#else
+#include <io.h>
+#endif
+
+#include <assert.h>
+#include <direct.h>
+#include <intrin.h>
+#include <malloc.h>
+#include <math.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <Windows.h>
+#include <WinSock2.h>
+
+#include <pthread.h>
+#include <pthread_clock.h>
+
+
+// Do nothing for the system log for now.
+inline void openlog(const char*, int, int) {}
+inline void closelog() {}
+inline void syslog(int, const char*, ...) {}
+
+#define SIGCHLD 0
+// We use SIGILL as SIGALRM, as SIGILL is
+// never generated by windows.
+#define SIGALRM SIGILL
+// All 4 of these are invalid signal IDs.
+#define SIGVTALRM 26
+#define SIGPROF 27
+#define SIGKILL 28
+#define SIGUSR1 29
+
+inline void __stdcall __alarm_callback_func(HWND, UINT, UINT_PTR, DWORD) {
+  raise(SIGALRM);
+}
+
+inline unsigned int alarm(unsigned int seconds) {
+  SetTimer(NULL, 0, (UINT)(seconds * 1000), __alarm_callback_func);
+  return 0;
+}
+
+inline char* asctime_r(const tm* tm, char* buf) {
+  char tmpBuf[64];
+  if (asctime_s(tmpBuf, sizeof(tmpBuf), tm))
```

Comment:
You shouldn't need to explicitly specify the buffer size,  because [Secure Template Overloads](https://msdn.microsoft.com/en-us/library/ms175759.aspx) allow for [template parameter inference](https://randomascii.wordpress.com/2013/04/03/stop-using-strncpy-already/), which is pretty cool.